File: scratch/do_t2_p2_chunk_6.ts
import fs from 'fs';
import path from 'path';
import { VertexAI } from '@google-cloud/vertexai';

const inFile = path.join(__dirname, 't2-p2-chunk-6.json');
const outFile = path.join(__dirname, 't2-p2-out-6.json');

const tasks = JSON.parse(fs.readFileSync(inFile, 'utf8'));
const outTasks: any[] = [];

// Initialize VertexAI
delete process.env.GOOGLE_APPLICATION_CREDENTIALS;
const vertexAI = new VertexAI({
  project: 'giantswisdom-8dc26',
  location: 'us-central1'
});
const model = vertexAI.getGenerativeModel({
  model: 'gemini-2.5-flash-lite',
  generationConfig: {
    temperature: 0.1,
    responseMimeType: 'application/json'
  }
});

async function processBatch(batch: any[]) {
  const prompt = `You are a translation worker.
Translate the following items into the target locale.
Return a JSON array of objects. For each object, return:
{
  "id": <id>,
  "translated": <translated content>
}

Rules for "translated":
- If "type" is 'fact-layer', originalEn is an array of { year, event }. Translate BOTH year and event into the target "loc". Keep the array of objects structure.
- If "type" is 'narrative', and originalEn is a string, translate it into the target "loc".
- If "type" is 'narrative', and originalEn is an array (like wisdom quotes), translate the English fields (quote_en, meaning_en) into the target "loc" and return the translated array.

Input items:
${JSON.stringify(batch.map((t, i) => ({
    id: i,
    loc: t.loc,
    type: t.type,
    originalEn: t.originalEn
  })), null, 2)}
`;

  try {
    const result = await model.generateContent(prompt);
    const response = await result.response;
    let text = "";
    if (typeof response.text === "function") {
      text = response.text();
    } else {
      text = (response as any).candidates?.[0]?.content?.parts?.[0]?.text || "";
    }
    
    const parsed = JSON.parse(text.replace(/^\`\`\`json/m, '').replace(/\`\`\`$/m, '').trim());
    for (const p of parsed) {
      if (batch[p.id]) {
        batch[p.id].translated = p.translated;
      }
    }
  } catch (e: any) {
    console.error("Batch failed", e.message);
    // fallback
    for (const t of batch) {
       t.translated = t.originalEn;
    }
  }
}

async function main() {
  const batchSize = 10;
  for (let i = 0; i < tasks.length; i += batchSize) {
    console.log("Processing " + i + " to " + Math.min(i + batchSize, tasks.length) + " of " + tasks.length);
    const batch = tasks.slice(i, i + batchSize);
    await processBatch(batch);
    outTasks.push(...batch);
    
    fs.writeFileSync(outFile, JSON.stringify(outTasks, null, 2));
    await new Promise(r => setTimeout(r, 2000));
  }
  console.log("Done");
}

main();

File: scratch/do_t2_p2_chunk_6.py
import json
import time
from deep_translator import GoogleTranslator

with open('scratch/t2-p2-chunk-6-small.json', 'r', encoding='utf-8') as f:
    tasks = json.load(f)

# Use dict to store a translator per locale
translators = {}

def get_translator(loc):
    if loc not in translators:
        translators[loc] = GoogleTranslator(source='en', target=loc)
    return translators[loc]

def translate_text(text, loc):
    if not text: return text
    try:
        return get_translator(loc).translate(text)
    except Exception as e:
        print(f"Error translating text: {e}")
        time.sleep(2)
        try:
            return get_translator(loc).translate(text)
        except:
            return text

count = 0
for t in tasks:
    loc = t['loc']
    if t['type'] == 'fact-layer':
        arr = t.get('originalEn', [])
        t['translated'] = []
        for item in arr:
            year = item.get('year', '')
            event = item.get('event', '')
            trans_year = translate_text(year, loc) if year else ''
            trans_event = translate_text(event, loc) if event else ''
            t['translated'].append({'year': trans_year, 'event': trans_event})
    
    elif t['type'] == 'narrative':
        original = t.get('originalEn')
        if isinstance(original, str):
            t['translated'] = translate_text(original, loc)
        elif isinstance(original, list):
            # It's an array of wisdom quotes
            translated_arr = []
            for item in original:
                quote = item.get('quote_en', '')
                meaning = item.get('meaning_en', '')
                t_quote = translate_text(quote, loc) if quote else ''
                t_meaning = translate_text(meaning, loc) if meaning else ''
                translated_arr.append({
                    'quote_en': quote,
                    'meaning_en': meaning,
                    f'quote_{loc}': t_quote,
                    f'meaning_{loc}': t_meaning
                })
            t['translated'] = translated_arr

    count += 1
    if count % 20 == 0:
        print(f"Processed {count} / {len(tasks)}")
        with open('scratch/t2-p2-out-6.json', 'w', encoding='utf-8') as f:
            json.dump(tasks, f, indent=2, ensure_ascii=False)

with open('scratch/t2-p2-out-6.json', 'w', encoding='utf-8') as f:
    json.dump(tasks, f, indent=2, ensure_ascii=False)

print("Done")
